External Data and Remotes (user-guide/external-data)

dashohoxha opened the pull request:
Close #520, Close #566, Close #237, Related #108

Consolidate (and deprecate):
- [External Dependencies](https://dvc.org/doc/user-guide/external-dependencies)
- [Managing External Data](https://dvc.org/doc/user-guide/managing-external-data)

Maybe the examples for each remote type (those that are collapsed) should be removed as well from these man pages:
- https://dvc.org/doc/command-reference/remote/add
- https://dvc.org/doc/command-reference/remote/modify

shcheklein commented:
@dashohoxha good stuff! Will keep reviewing and asking questions.

First is - what is the difference between "Create Remotes" and "DVC storage" sections?

shcheklein commented:
@dashohoxha I also not sure how we went from solving #520 to actually including regular remotes stuff? #520 is about very specific advanced thing - managing external data, while remotes is a very basic functionality. It deserves it's own very details section that explains them very well.

dashohoxha commented:
> what is the difference between "Create Remotes" and "DVC storage" sections

"Creates Remotes" explains all the details and options about a certain remote type (which are set with `dvc remote modify`).

"DVC storage" emphasizes that we need to use the option `-d, --default` so that the commands `dvc push` and `dvc pull` can use that remote as a "mirror" of  `.dvc/cache` (aka a DVC Storage).

I admit that in some cases there is not much difference between them and they look very similar.

> I also not sure how we went from solving #520 to actually including regular remotes stuff? #520 is about very specific advanced thing - managing external data, while remotes is a very basic functionality. It deserves it's own very details section that explains them very well.

External data and remotes seem to me closely related because:
1. We can use remotes to access and manage external data.
2. External data management is specific for each remote type. For example some remote types can be used for external outputs, some others cannot.

So, I thought that it is best to explain them together.

shcheklein commented:
> I admit that in some cases there is not much difference between them and they look very similar.

let's remove and simplify

> External data and remotes seem to me closely related because:

introducing basic and advanced concepts together sends a wrong signal. We should not emphasize external deps/outs as a regular basic workflow. Also, there should be a separate motivation behind them.

dashohoxha commented:
> > I admit that in some cases there is not much difference between them and they look very similar.
> 
> let's remove and simplify

I moved the sections about DVC Storages as collapsed sections under "Remotes".

> introducing basic and advanced concepts together sends a wrong signal. We should not emphasize external deps/outs as a regular basic workflow. Also, there should be a separate motivation behind them.

Are you suggesting to split Remotes from External Data, like this:

- user-guide/remotes/
  - local
  - http
  - ssh
  - amazon
  - s3
  - azure
  - gs
  - hdfs
  - oss
- user-guide/external-data/
  - local
  - http
  - ssh
  - amazon
  - gs
  - hdfs

shcheklein commented:
> I moved the sections about DVC Storages as collapsed sections under "Remotes".

I just don't get the purpose of those sections. Can you elaborate or explain it in some other way? Why do we need both - remotes and dvc storage?

> Are you suggesting to split Remotes from External Data, like this:

Yes. I'm not sure about exact place and structure. But external data management should be separate for sure. Remotes - they can be part of the Basic Concepts, they can be part of the Managing Data section. External Data Management should be part of the Advanced section of the User Guide. 

^^ this is not a precise instruction, it's just a way for me to communicate how far I think those two should be split.

dashohoxha commented:
> let's remove and simplify

I removed the sections about "DVC storage".

I also split the "Remotes" from "External Data":
- https://dvc-org-pr-807.herokuapp.com/doc/user-guide/remotes
- https://dvc-org-pr-807.herokuapp.com/doc/user-guide/external-data

The first one ("Remotes") is basically a consolidation of remote types on:
- https://dvc.org/doc/command-reference/remote/add#supported-storage-types
- https://dvc.org/doc/command-reference/remote/modify#available-settings-per-storage-type

I seems like the command `dvc remote` only takes care of the configuration file `.dvc/config` (or `.dvc/config.local` etc.) One could modify these configuration files manually (using an editor, without using `dvc remote`) and everything would be fine. So, it seems more appropriate to explain different remote types on the User Guide.

The second section ("External Data") is an attempt to consolidate and improve these pages:
- https://dvc.org/doc/user-guide/external-dependencies
- https://dvc.org/doc/user-guide/managing-external-data

dashohoxha commented:
I am not sure that I understand this note from [managing-external-data](https://dvc.org/doc/user-guide/managing-external-data):

> Avoid using the same remote location that you are using for `dvc push`, `dvc pull`, `dvc fetch` as external cache for your external outputs, because it may cause possible checksum overlaps. Checksum for some data file on an external storage can potentially collide with checksum generated locally for a different file, with a different content.

The first part seems reasonable: "Avoid using the same remote location".

The explanation seems nonsense. If there is some possibility for checksum collision, this may happen for the cached files as well (two different cached files happen to have the same checksum). If those external data files that we are tracking with DVC were instead local files, we would store them in cache with all the other project files. So, what difference does it make if they are tracked remotely?

Maybe some other explanation is needed or maybe we should remove this note.

shcheklein commented on src/Documentation/sidebar.json:
```diff
@@ -120,6 +120,80 @@
+        ]
+      },
+      {
+        "label": "External Data",
```
why can't we keep the Managing External Data label and name? Not much difference but redirect won't be required.

dashohoxha replied:
> why can't we keep the Managing External Data label and name? Not much difference but redirect won't be required.

Redirect would still be required because of "external-dependencies".
On the other hand "External Data" is more generic than "Managing External Data". For example in the case of HTTP (and other cases) we are just accessing or downloading external data, "managing" doesn't seem suitable.

shcheklein replied:
Again, please, don't resolve. I have to repeat myself and create duplicates.

Don't see any reasons to rename and complicate this.

dashohoxha replied:
> Don't see any reasons to rename and complicate this.

I don't understand, what is the complication that comes from renaming? In the first comment you say "not much difference", and I agree with this. And redirect is still required even if we don't change the name.

shcheklein replied:
one extra redirect to support + updating links in the docs

dashohoxha replied:
I have already updated the links on the docs, and also have fixed the redirect.
Changing the name back would be extra work, and unjustified, unless the old name is better than the new one.

shcheklein commented on src/Documentation/sidebar.json:
```diff
@@ -120,6 +120,80 @@
         "slug": "managing-external-data"
       },
+      {
+        "label": "Remotes",
```
Come up with a better name - Setting up storage?

Move the section up.

dashohoxha replied:
> Come up with a better name - Setting up storage?

I am renaming to "DVC Remotes". I am using "DVC Storage" for the mirror of `.dvc/cache/`. A "DVC Remote" is analogous to a Git Remote, and is broader than a "DVC Storage" (which is just a mirror/backup of the local cache). This terminology is just a convention (for example instead of "DVC Storage" we could have used "DVC Cache Backup".

shcheklein replied:
Please don't resolve comments that are not created by you. As I mentioned already, it's extremely hard to follow and track changes this way.

I would prefer to have something human readable . Otherwise we are defining remote through remotes. `Remote is remote ...` kind-a of thing.

dashohoxha replied:
> I would prefer to have something human readable . Otherwise we are defining remote through remotes. Remote is remote ... kind-a of thing.

I don't understand this. We are talking about the remotes that are managed with the command `dvc remote`. What could be a better name than "Remote" or "DVC Remote"?

shcheklein commented on src/Documentation/sidebar.json:
```diff
@@ -120,6 +120,80 @@
         "label": "Managing External Data",
```
remove this section, docs as part of this PR if you intend to replace. It's fine to do this when we finalize the structure first, but should be removed/replaced as part of this PR.

dashohoxha replied:
> remove this section

I am removing it.

shcheklein commented on src/Documentation/sidebar.json:
```diff
@@ -120,6 +120,80 @@
+        "source": "external-data/index.md",
+        "children": [
+          {
+            "label": "Local Filesystem",
```
Local should go very last in all cases. SSH,  S3, Google Cloud, Azure, HTTP, HDFS - something like this is better

dashohoxha replied:
Local supports more cases than the other types, and it also can be used for the mounted data storages (NAS etc.) So, it seems like an important one. Why should it go last?

shcheklein commented on static/docs/user-guide/external-data/amazon.md:
```diff
@@ -0,0 +1,77 @@
+# External Data on Amazon S3
+
+Amazon S3 or Amazon Simple Storage Service is a service offered by Amazon Web
```
let's remove this, no need to explain this

shcheklein replied:
better give an overview - we support both dependencies and outputs natively, mention important caveats if any etc ... applies to other storage types

dashohoxha replied:
> let's remove this, no need to explain this

Agree

shcheklein commented on static/docs/user-guide/external-data/amazon.md:
````diff
@@ -0,0 +1,77 @@
+$ dvc import-url s3://mybucket/file.csv
+```
+
+Or, using a remote:
````
remote -> remote notation?

shcheklein replied:
or put `remote`

shcheklein replied:
it's kinda not only remote as usual it's also a protocol, so some explanation is needed

dashohoxha replied:
> remote -> remote notation?

First we create a remote (with `dvc remote add`), so "remote" is ok.
I think that there is no need to introduce the term "remote notation", because it is kind of obvious.

shcheklein commented on static/docs/user-guide/external-data/amazon.md:
```diff
@@ -0,0 +1,77 @@
+
+## External Data and Outputs
+
+For cached external outputs (specified using `-o`) we need to setup an external
```
`dvc add` creates an output as well

dashohoxha replied:
Amended.

shcheklein commented on static/docs/user-guide/external-data/gs.md:
```diff
@@ -0,0 +1,77 @@
+
+### Warning: Don't use the same location for the external cache and for the DVC storage
+
+When you setup an external cache for your external outputs, avoid using the same
```
I don't like that we repeat over and over again. It feels like we can put into index page, general overview page or something?

shcheklein commented on static/docs/user-guide/external-data/hdfs.md:
```diff
@@ -0,0 +1,89 @@
+# External Data on HDFS
```
let's remove all external data section for HDFS. We don't support it actually and it should be remove from DVC for now. cc @efiop can confirm that.

efiop replied:
Confirmed, we were not able to find a suitable workaround for external outputs. 🙁 So let's hide hdfs external deps and outs from docs for now. We'll probably remove them from dvc too very soon.

dashohoxha replied:
> So let's hide hdfs external deps and outs from docs for now.

I am removing the section "External Data and Outputs". But it seems like "External Dependencies" should work. Please correct me if I am wrong.

shcheklein replied:
@dashohoxha even external dependencies do not work reliable unfortunately as far as I remember

dashohoxha replied:
Removed.

shcheklein commented on static/docs/user-guide/external-data/http.md:
```diff
@@ -0,0 +1,66 @@
+# External Data on HTTP/HTTPS
+
+**❗ Note:** HTTP/HTTPS remotes support only download operations.
```
it means probably that we support only dependencies, right? use the same terminology ... not download operations

dashohoxha replied:
> it means probably that we support only dependencies, right? use the same terminology ... not download operations

I think it is better to remove this note, since it says the obvious (there is only the section for dependencies, no section for outputs).

shcheklein replied:
We've seen people asking us over and over again - thus this note. It was not cleat either docs are out of date (always) or what :)

dashohoxha replied:
Maybe it was useful when HTTP was in the same page with the other remotes, but in this page it does not make sense.
Probably people are confused by Git remotes, which can be used both to pull and push.

shcheklein commented on static/docs/user-guide/external-data/index.md:
```diff
@@ -0,0 +1,22 @@
+# External Data
+
+External data are located outside the project directory, either on the local
```
I think we had better explanation before why and when it should be used.

dashohoxha replied:
Do you mean this explanation:

> There are cases when data is so large, or its processing is organized in a way that you would like to avoid moving it out of its external/remote location. For example from a network attached storage (NAS) drive, processing data on HDFS, running Dask via SSH, or having a script that streams data from S3 to process it. A mechanism for external outputs and external dependencies provides a way for DVC to control data externally.

shcheklein replied:
Yes.

shcheklein commented on static/docs/user-guide/external-data/index.md:
```diff
@@ -0,0 +1,22 @@
+# External Data
+
+External data are located outside the project directory, either on the local
+filesystem or on a remote host. They may be used in order to:
```
remote host sounds like a machine while it can be actually a cloud

dashohoxha replied:
> remote host sounds like a machine while it can be actually a cloud

"remote host" (like a machine) is one the cases. I am also adding "cloud data storage", like this:

> External data are located outside the project directory, either on the local
filesystem, or on a remote host, or on a cloud data storage.

shcheklein commented on static/docs/user-guide/external-data/local.md:
```diff
@@ -0,0 +1,311 @@
+
+</details>
+
+## External Deps and Outs
```
dependencies and outputs

shcheklein commented on static/docs/user-guide/external-data/local.md:
```diff
@@ -0,0 +1,311 @@
+# External Data on a Local Filesystem
+
+Sometimes we would like to keep the data and cache outside the project
```
We should mention that it's not a default and recommended way. The recommended setup is the shared cache in this case.

dashohoxha replied:
> We should mention that it's not a default and recommended way. The recommended setup is the shared cache in this case.

The word "Sometimes" already implies that this is not usual. By the way, this case is not related to the case of "shared cache".

shcheklein commented on static/docs/user-guide/external-data/local.md:
```diff
@@ -0,0 +1,311 @@
+[network-attached storage (NAS)](https://searchstorage.techtarget.com/definition/network-attached-storage)
+and is mounted through NFS, SSHFS, Samba, etc.
+
+## Tracking External Data
```
It should be using the same structure - dependencies, outputs. Or other sections should be changed.

dashohoxha replied:
> It should be using the same structure

It is not possible to use the same structure because the "local filesystem" is different from the other types of external data.

shcheklein replied:
why is it different?

dashohoxha replied:
1. With "local" you can have an external cache (used instead of `.dvc/cache/`), with the other types this is not possible.
2. "local" is available to all the users, the other types might not be available or usually is available only one.

shcheklein commented on static/docs/user-guide/external-data/local.md:
```diff
@@ -0,0 +1,311 @@
+
+<details>
+
+### Using an environment variable
```
these hack apply to any storage type, I would generalize and move them somewhere ... not sure actually where and how yet

dashohoxha replied:
> these hack apply to any storage type, I would generalize and move them somewhere ... not sure actually where and how yet

You are right, it applies to the other storage types too, but I don't see a better place either. The "index" page would not be a good place because you need to explain first the normal way, before explaining the hack (or the tip, or the optimization). Let's leave it on the local page and the user can generalize it for the other storage types himself.

By the way, the "local" storage type applies to all the users, so all of them need to read it. The other storage types don't apply (or are not relevant) to all the users. For example a user may use S3, but not GS etc.

shcheklein replied:
My point is that these pages should be very simple. Most of the information should be generalized and provided a two pages - external dependencies and external outputs.

> Let's leave it on the local page and the user can generalize it for the other storage types himself.

I don't see this as a reasonable solution.

> By the way, the "local" storage type applies to all the users

May be I'm not following something, but no, external "local" data management is a very specific case. Mostly happening due to misunderstanding of the DVC default setup - shared cache.

dashohoxha replied:
"shared cache" itself is a case of "local" external data (since the cache is outside the project directory).

dashohoxha commented:
I would appreciate if @jorgeorpinel could help with improving this PR, since he is much better than me for making fine-tuned changes.

shcheklein commented:
@dashohoxha it can require some significant changes. And to understand the logic, clarify with you and implement would require a lot of time. Considering that @jorgeorpinel has his own queue of features and PRs it will take a lot of time before we get there unfortunately.

shcheklein commented on static/docs/user-guide/remotes/ssh.md:
```diff
@@ -0,0 +1,177 @@
+
+<details>
+
+### Tip: Recommended SSH configuration
```
just put an external link, not need to provide the full instruction on how to use SSH keys. Also setup looks too complicated, you would usually just add your pub key to the `authorized_keys` file on the server.

dashohoxha replied:
No external site would explain it simpler than this. And it is not just sending the pub key to the server, the configuration of ssh (on `~/.ssh/config`) also matters for simplicity. It is thanks to this config file that we don't need to tell DVC neither the username, nor the key file or the port.
So this section is tuned to DVC, and no external site wold explain SSH configuration like this.

shcheklein replied:
>  It is thanks to this config file that we don't need to tell DVC neither the username, nor the key file or the port.

that's the point, it's again very advanced very complicated workflow you are describing. You don't use it usually. In 99% you just add your public key to the server. And ssh client is using it by default (I hope that paramiko that we use internally is set to do the regular stuff).

> So this section is tuned to DVC, and no external site wold explain SSH configuration like this.

by default not that much special about DVC and it's enough to name a few options to override key, username, etc.

dashohoxha replied:
> In 99% you just add your public key to the server.

Using the same key for all the servers is not considered a good practice.
I use different keys for different servers, and even for different services in the same server.

> ssh client is using it by default

ssh client is using it automatically if you tell it in the configuration file (`.ssh/config`) which key to use.

shcheklein replied:
@dashohoxha this is the link https://www.digitalocean.com/community/tutorials/how-to-set-up-ssh-keys--2 that should be more than enough. To be honest I'm not even sure if we need even the link. 

> Using the same key for all the servers is not considered a good practice.

one encrypted key is very regular and reasonably secure. Please, share some links that prove your point, similar to one above.

> ssh client is using it automatically if you tell it in the configuration file (.ssh/config) which key to use.

I'm not sure about specific mechanics but I don't remember the last time I needed to make any changes to the config. Again, would love some links with the recommendations on what I am doing wrong,

dashohoxha replied:
What is the usefulness of recommending a link that everybody can google and find? Besides, it is non-authoritative and already too old.

> one encrypted key is very regular and reasonably secure

This is like using the same password for all the sites.

> Please, share some links that prove your point, similar to one above.

The link that you provided does not prove anything. The links that I may provide do not prove anything. But some reasons why someone might want to use more than a key are on these pages:

- https://www.redhat.com/sysadmin/manage-multiple-ssh-key-pairs
- https://guide.freecodecamp.org/ssh/managing-multiple-ssh-keys/
- https://security.stackexchange.com/questions/40050/best-practice-separate-ssh-key-per-host-and-user-vs-one-ssh-key-for-all-hos

shcheklein replied:
> What is the usefulness of recommending a link that everybody can google and find?

Good question. That's exactly the point. No need to include all this information at all. Just mention that we don't recommend using plain text passwords.

> This is like using the same password for all the sites.

Comparing passwords and keys is wrong. They have very different security characteristics.

```
As for universally-applicable guidelines on how to run your security: there are none. The more additional security you add, the more convenience you give up. The one piece of advice I can give categorically is this: keep your private key encrypted. The added security there is pretty significant.
```

Good points.

shcheklein commented on static/docs/user-guide/remotes/ssh.md:
```diff
@@ -0,0 +1,177 @@
+
+<details>
+
+### Details: All remote options
```
Let's not hide it. It should be similar to command reference sections - same structure for all pages, list of options, description with notes, stc

dashohoxha replied:
> Let's not hide it.

Good point. I don't know why I have made it collapsed.

> It should be similar to command reference sections

It is actually a copy/paste from the command reference sections

By the way, should we keep this information duplicated (here and on the command reference), or should we remove it from the command reference?

shcheklein replied:
> By the way, should we keep this information duplicated (here and on the command reference), or should we remove it from the command reference?

Good question. I don't have a strong opinion, but it's probably better to simplify the command reference and remove duplicates.

dashohoxha replied:
Done.

shcheklein commented on src/Documentation/sidebar.json:
```diff
@@ -115,10 +115,79 @@
       {
-        "label": "Managing External Data",
-        "slug": "managing-external-data"
+        "label": "DVC Remotes",
```
let's split remotes into a separate PR. I don't have capacity to review a few PRs of 2000+ lines of code. At this pace of addressing comments they won't even be merged :(

dashohoxha replied:
Let's keep them in the same PR but focus the review on External Data first, and after we are done with that focus on Remotes.
By splitting them we would lose context and discussions, and it is also a lot of work to undo and redo.

shcheklein commented on src/Documentation/sidebar.json:
```diff
@@ -115,10 +115,79 @@
       "large-dataset-optimization",
-      "external-dependencies",
       {
-        "label": "Managing External Data",
```
I still don't see a single reason to do this. HTTP is an edge case that should hardly affect this decision.

shcheklein commented on static/docs/command-reference/import-url.md:
```diff
@@ -79,8 +78,8 @@ DVC supports several types of (local or) remote locations (protocols):
 
 Another way to understand the `dvc import-url` command is as a short-cut for a
 more verbose `dvc run` command. This is discussed in the
-[External Dependencies](/doc/user-guide/external-dependencies) documentation,
```
I think we should have external dependencies and external outputs sections still. They both should be introductory, provide some context about each.

dashohoxha replied:
I don't understand this, what do you mean?

shcheklein commented on static/docs/install/linux.md:
````diff
@@ -12,10 +12,9 @@
 ```
 
-Depending on the type of the
-[remote storage](/doc/user-guide/external-dependencies) you plan to use, you
````
`/doc/user-guide/external-dependencies`  is a bug! I'm not sure why do we have this link at all here.

shcheklein replied:
I'm fixing this right away ... no reason to wait for fixes like this

shcheklein replied:
Fixed by https://github.com/iterative/dvc.org/commit/50bc171793b31b6d94c35baaccf7e348e8e096a7

dashohoxha replied:
You have fixed it in a way that creates conflicts, not as part of this PR.

shcheklein replied:
Yep. Resolving conflicts is a very regular part of the modern dev process. That's why git has rebase, merge, etc, etc. Even Github can resolve automatically majority of conflicts.

We can't wait too long to merge this to fix the existing bug.

shcheklein commented:
Again, too complicated, too many changes at once, no luck of iterating on this not willingness to address even small changes.
